File: frontend/uks/src/services/repo/repo.service.ts
import { RepoForkRequest, RepoUserRequest, RepoUpdateRequest, WatchStarResponseDTO } from 'src/models/repo/repo';
import {
  EditRepoRequest,
  RepoBasicInfoDTO,
  RepoRequest,
} from 'src/models/repo/repo';
import { Observable, of } from 'rxjs';
import { environment } from 'src/environments/environment';
import { UserBasicInfo } from 'src/models/user/user';
import { HttpRequestService } from 'src/utils/http-request.service';

import { Injectable } from '@angular/core';
import { BranchDTO } from 'src/models/branch/branch';

@Injectable({
  providedIn: 'root',
})
export class RepoService {


  constructor(
    private httpRequestService: HttpRequestService,
  ) { }

  getAllPublic(): Observable<RepoBasicInfoDTO[]> {
    const url = environment.API_BASE_URL + '/repo/getAllPublic';
    return this.httpRequestService.get(url) as Observable<RepoBasicInfoDTO[]>;
  }

  getMyRepos(userId: string): Observable<RepoBasicInfoDTO[]> {
    const url = environment.API_BASE_URL + `/repo/getMyRepos/${userId}`;
    return this.httpRequestService.get(url) as Observable<RepoBasicInfoDTO[]>;
  }
  getById(repoId: string): Observable<RepoBasicInfoDTO | null> {
    const url = environment.API_BASE_URL + `/repo/getById/${repoId}`;
    return this.httpRequestService.get(url) as Observable<RepoBasicInfoDTO | null>;
  }


  createNewRepo(repoRequest: RepoRequest): Observable<RepoBasicInfoDTO | null> {
    const url = environment.API_BASE_URL + '/repo/create';
    const body = JSON.stringify(repoRequest);

    return this.httpRequestService.post(
      url,
      body
    ) as Observable<RepoBasicInfoDTO | null>;
  }

  updateRepo(
    id: string,
    request: RepoUpdateRequest
  ): Observable<RepoBasicInfoDTO | null> {
    const url = environment.API_BASE_URL + `/repo/update/${id}`;
    const body = JSON.stringify(request);

    return this.httpRequestService.put(
      url,
      body
    ) as Observable<RepoBasicInfoDTO | null>;
  }

  validateOverviewByRepoName(
    repoRequest: RepoRequest
  ): Observable<RepoBasicInfoDTO | null> {
    const url = environment.API_BASE_URL + '/repo/validateOverviewByRepoName';
    const body = JSON.stringify(repoRequest);

    return this.httpRequestService.post(
      url,
      body
    ) as Observable<RepoBasicInfoDTO | null>;
  }

  canEditRepoItems(repoRequest: EditRepoRequest): Observable<boolean> {
    const url = environment.API_BASE_URL + '/repo/canEditRepoItems';
    const body = JSON.stringify(repoRequest);

    return this.httpRequestService.post(url, body) as Observable<boolean>;
  }

  getCanEditRepoItems(): boolean {
    var storedBoolean = localStorage.getItem('canEditRepoItems');
    return storedBoolean === 'true' ? true : false;
  }

  getAllForked(repoId: string) {
    const url = environment.API_BASE_URL + `/repo/getAllForked/${repoId}`;
    return this.httpRequestService.get(url) as Observable<RepoBasicInfoDTO[]>;
  }

  newFork(forkRequest: RepoForkRequest) {
    const url = environment.API_BASE_URL + '/repo/fork';
    const body = JSON.stringify(forkRequest);

    return this.httpRequestService.post(url, body) as Observable<RepoBasicInfoDTO | null>;
  }


  star(starRequest: RepoUserRequest) {
    const url = environment.API_BASE_URL + '/repo/star';
    const body = JSON.stringify(starRequest);

    return this.httpRequestService.post(url, body) as Observable<void>;
  }
  getAllStargazers(repoId: string) {
    const url = environment.API_BASE_URL + `/repo/getAllStargazers/${repoId}`;
    return this.httpRequestService.get(url) as Observable<UserBasicInfo[]>;
  }

  watch(watchRequest: RepoUserRequest) {
    const url = environment.API_BASE_URL + '/repo/watch';
    const body = JSON.stringify(watchRequest);

    return this.httpRequestService.post(url, body) as Observable<void>;
  }
  getAllWatchers(repoId: string) {
    const url = environment.API_BASE_URL + `/repo/getAllWatchers/${repoId}`;
    return this.httpRequestService.get(url) as Observable<UserBasicInfo[]>;
  }

  amIWatchingStargazing(request: RepoUserRequest) {
    const url = environment.API_BASE_URL + `/repo/amIWatchingStargazing`;
    const body = JSON.stringify(request);

    return this.httpRequestService.post(url, body) as Observable<WatchStarResponseDTO>;
  }

  deleteRepo(repoId: string) {
    const url = environment.API_BASE_URL + `/repo/delete/${repoId}`;
    return this.httpRequestService.delete(url) as Observable<void>;
  }

  getDefaultBranch(repoId: string): Observable<BranchDTO> {
    const url = environment.API_BASE_URL + `/repo/getDefaultBranch/${repoId}`;
    return this.httpRequestService.get(url) as Observable<BranchDTO>;
  }
}

File: backend/uks/src/main/java/uns/ac/rs/uks/service/RepoService.java
package uns.ac.rs.uks.service;

import jakarta.persistence.EntityManager;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import uns.ac.rs.uks.dto.request.*;
import uns.ac.rs.uks.dto.response.BranchDTO;
import uns.ac.rs.uks.dto.response.RepoBasicInfoDTO;
import uns.ac.rs.uks.dto.response.UserDTO;
import uns.ac.rs.uks.dto.response.WatchStarResponseDTO;
import uns.ac.rs.uks.exception.NotAllowedException;
import uns.ac.rs.uks.exception.NotFoundException;
import uns.ac.rs.uks.mapper.BranchMapper;
import uns.ac.rs.uks.mapper.RepoMapper;
import uns.ac.rs.uks.mapper.UserMapper;
import uns.ac.rs.uks.model.*;
import uns.ac.rs.uks.repository.repo.RepoRepository;

import java.util.List;
import java.util.UUID;

@Service
public class RepoService {
    @Autowired
    private RepoRepository repoRepository;
    @Autowired
    private UserService userService;
    @Autowired
    private MemberService memberService;
    @Autowired
    private BranchService branchService;
    @Autowired
    private EntityManager entityManager;
    @Autowired
    private GitoliteService gitoliteService;

//    @Cacheable(value = "repos")
    public List<RepoBasicInfoDTO> getAllPublic() {
        List<Repo> allPublic = repoRepository.findAllByIsPublicTrue();
        return RepoMapper.toDTOs(allPublic);
    }

    @Cacheable(value = "repos", key = "#userID")
    public List<RepoBasicInfoDTO> getMyRepos(UUID userID) {
        List<Repo> allPublic = repoRepository.findAllByOwnerId(userID);
        return RepoMapper.toDTOs(allPublic);
    }

    @CacheEvict(value = "repos", allEntries = true)
    public RepoBasicInfoDTO createNewRepo(RepoRequest repoRequest) throws NotFoundException {
        User user = userService.getById(repoRequest.getOwnerId());
        Repo repo = RepoMapper.toRepoFromRequest(repoRequest, user);
        repo.setDefaultBranch(branchService.createDefaultBranch(repo, user));

        repo.setCloneUri(gitoliteService.createRepo(repo.getName(), user.getUsername()));
        memberService.addNewMember(user, repo, RepositoryRole.OWNER, MemberInviteStatus.ACCEPTED);

        return RepoMapper.toDTO(repoRepository.save(repo));
    }

    public Repo getById(UUID repoId) {
        return repoRepository.findById(repoId).orElseThrow(() -> new NotFoundException("Repository not found."));
    }
    public RepoBasicInfoDTO findById(UUID repoId) {
        return RepoMapper.toDTO(getById(repoId));
    }

    public RepoBasicInfoDTO getByNameAndPublicOrMember(RepoRequest repoRequest) {
        List<Repo> repos = repoRepository.findAllByName(repoRequest.getName());
        for (Repo repo : repos) {
            if (repo.getIsPublic()) {
                return RepoMapper.toDTO(repo);
            }
            Member member = memberService.findMemberByUserIdAndRepositoryId(repoRequest.getOwnerId(), repo.getId());
            if (member != null) {
                return RepoMapper.toDTO(repo);
            }
        }
        return null;
    }

    public Boolean canEditRepoItems(EditRepoRequest repoRequest) {
        Member member = memberService.findMemberByUserIdAndRepositoryId(repoRequest.getUserId(),
                repoRequest.getRepoId());
        return member != null;
    }

    @CacheEvict(value = "repos", allEntries = true)
    public RepoBasicInfoDTO updateRepo(UUID repositoryId, RepoUpdateRequest request) {
        Repo repo = getById(repositoryId);
        repo.setName(request.getName());
        repo.setDescription(request.getDescription());
        repo.setIsPublic(request.getIsPublic());
        repo.setDefaultBranch(branchService.getById(request.getDefaultBranch()));

        repoRepository.save(repo);
        return RepoMapper.toDTO(repo);
    }

    @CacheEvict(value = "repos", allEntries = true)
    public RepoBasicInfoDTO forkRepo(RepoForkRequest forkRequest) {
        checkForkValidity(forkRequest);
        User owner = entityManager.getReference(User.class, forkRequest.getOwnerId());
        Repo repo = getById(forkRequest.getOriginalRepoId());

        Repo newRepo = RepoMapper.map(repo);
        newRepo.setOwner(owner);
        newRepo.setName(forkRequest.getName());
        newRepo.setDescription(forkRequest.getDescription());
        newRepo.setForkParent(repo);
        newRepo.setCloneUri(gitoliteService.createRepo(newRepo.getName(), owner.getUsername()));

        return RepoMapper.toDTO(repoRepository.save(newRepo));
    }

    private void checkForkValidity(RepoForkRequest forkRequest) {
        if(!forkRequest.getIsPublic()){
            Member member = memberService.findMemberByUserIdAndRepositoryId(
                    forkRequest.getOwnerId(), forkRequest.getOriginalRepoId());
            if(member == null) {
                throw new NotAllowedException("Fork not allowed");
            }
        }
    }

    public List<RepoBasicInfoDTO> getAllForked(UUID repoId) {
        Repo repo = getById(repoId);
        return RepoMapper.toDTOs(repo.getForkChildren());
    }

    public RepoBasicInfoDTO starRepo(RepoUserRequest starRequest) {
        Repo repo = getById(starRequest.getRepoId());
        User user = userService.getById(starRequest.getUserId());
        boolean isStargazing = repo.getStaredBy().stream().anyMatch(u -> u.getId().equals(user.getId()));
        if(!isStargazing) {
            user.getStared().add(repo);
            repo.getStaredBy().add(user);
        } else {
            user.getStared().remove(repo);
            repo.getStaredBy().remove(user);
        }
        return RepoMapper.toDTO(repoRepository.save(repo));
    }

    public RepoBasicInfoDTO watchRepo(RepoUserRequest starRequest) {
        Repo repo = getById(starRequest.getRepoId());
        User user = userService.getById(starRequest.getUserId());
        boolean isWatching = repo.getWatchers().stream().anyMatch(u -> u.getId().equals(starRequest.getUserId()));
        if(!isWatching) {
            user.getWatching().add(repo);
            repo.getWatchers().add(user);
        } else {
            user.getWatching().remove(repo);
            repo.getWatchers().remove(user);
        }
        return RepoMapper.toDTO(repoRepository.save(repo));
    }

    public List<UserDTO> getAllStargazers(UUID repoId) {
        Repo repo = getById(repoId);
        return UserMapper.toDTOs(repo.getStaredBy());
    }

    public List<UserDTO> getAllWatchers(UUID repoId) {
        Repo repo = getById(repoId);
        return UserMapper.toDTOs(repo.getWatchers());
    }

    public WatchStarResponseDTO amIWatchingStargazing(RepoUserRequest request) {
        Repo repo = getById(request.getRepoId());
        boolean isWatching = repo.getWatchers().stream().anyMatch(u -> u.getId().equals(request.getUserId()));
        boolean isStargazing = repo.getStaredBy().stream().anyMatch(u -> u.getId().equals(request.getUserId()));
        return new WatchStarResponseDTO(isWatching, isStargazing);
    }

    @CacheEvict(value = "repos", allEntries = true)
    public void deleteRepo(UUID repoId) {
        memberService.removeAllMembersFromRepo(repoId);
        repoRepository.deleteById(repoId);
    }

    public BranchDTO getDefaultBranch(UUID repoId) {
        Repo repo = getById(repoId);
        return BranchMapper.toDTO(repo.getDefaultBranch());
    }
}
